ee/modules/110-istio/hooks/dataplane_metadata_exporter.go

```diff
@@ -174,57 +249,205 @@ func applyIstioPodFilter(obj *unstructured.Unstructured) (go_hook.FilterResult,
+
 	globalRevision := versionMap[input.Values.Get("istio.internal.globalVersion").String()].Revision
 
 	input.MetricsCollector.Expire(metadataExporterMetricsGroup)
 
-	var namespaceRevisionMap = map[string]string{}
+	istioNamespaceMap := make(map[string]IstioNamespaceFilterResult)
 	for _, ns := range append(input.Snapshots["namespaces_definite_revision"], input.Snapshots["namespaces_global_revision"]...) {
-		nsInfo := ns.(NamespaceInfo)
-		if nsInfo.Revision == "global" {
-			namespaceRevisionMap[nsInfo.Name] = globalRevision
+		nsInfo := ns.(IstioNamespaceFilterResult)
+		if nsInfo.RevisionRaw == "global" {
+			nsInfo.Revision = globalRevision
 		} else {
-			namespaceRevisionMap[nsInfo.Name] = nsInfo.Revision
+			nsInfo.Revision = nsInfo.RevisionRaw
+		}
+		istioNamespaceMap[nsInfo.Name] = nsInfo
+	}
+
+	// upgradeCandidates[kind][namespace][name]Candidate{}
+	upgradeCandidates := make(map[string]map[string]map[string]Candidate)
+
+	k8sControllers := make([]go_hook.FilterResult, 0)
+	k8sControllers = append(k8sControllers, input.Snapshots["deployment"]...)
+	k8sControllers = append(k8sControllers, input.Snapshots["statefulset"]...)
+	k8sControllers = append(k8sControllers, input.Snapshots["daemonset"]...)
+
+	for _, k8sControllerRaw := range k8sControllers {
+		k8sController := k8sControllerRaw.(K8SControllerFilterResult)
+
+		// check if AutoUpgrade Label Exists on namespace
+		var namespaceAutoUpgradeLabelExists bool
+		if k8sControllerNS, ok := istioNamespaceMap[k8sController.Namespace]; ok {
+			namespaceAutoUpgradeLabelExists = k8sControllerNS.AutoUpgradeLabelExists
+		}
+
+		// if an istio.deckhouse.io/auto-upgrade Label exists in the namespace or in the controller
+		// and the controller is available for upgrade -> add to upgradeCandidates map
+		if (namespaceAutoUpgradeLabelExists || k8sController.AutoUpgradeLabelExists) && k8sController.AvailableForUpgrade {
+			if _, ok := upgradeCandidates[k8sController.Kind]; !ok {
+				upgradeCandidates[k8sController.Kind] = make(map[string]map[string]Candidate)
+			}
+			if _, ok := upgradeCandidates[k8sController.Namespace]; !ok {
+				upgradeCandidates[k8sController.Kind][k8sController.Namespace] = make(map[string]Candidate)
+			}
+			upgradeCandidates[k8sController.Kind][k8sController.Namespace][k8sController.Name] = Candidate{}
+		}
+	}
+
+	// replicaSets[namespace][replicaset-name]owner
+	replicaSets := make(map[string]map[string]Owner)
+
+	// create a map of the replica sets depending on the deployments from upgradeCandidates map
+	for _, rs := range input.Snapshots["replicaset"] {
+		rsInfo := rs.(K8SControllerFilterResult)
+		if rsInfo.Owner.Kind == "Deployment" {
+			if _, ok := upgradeCandidates["Deployment"][rsInfo.Namespace][rsInfo.Owner.Name]; ok {
+				if _, ok := replicaSets[rsInfo.Namespace]; !ok {
+					replicaSets[rsInfo.Namespace] = make(map[string]Owner)
+				}
+				replicaSets[rsInfo.Namespace][rsInfo.Name] = Owner{
```

Comment:
It could be a pointer to the candidate.

Reply:
Yes, it could be, but the logic for finding the owner is in a different block